### ISSUE TITLE
Concurrency Models

### ISSUE THREAD

**Issue by hij1nx** [jamii/make-you-a-database#1]
Relational systems are simple until you add concurrency at volume ;) Do you plan to use multi-version concurrency control or some other approach? Imho, say it's less about historic baggage than the nature of the problem. There are lots of brand new relational databases that are incredibly complex because mutable data structures and concurrency combine to create hard problems, especially when you start trying to determine which parts of ACID you care about.

Anyway, it's roughly like what Leslie Lamport: If you try to create a distributed, fault tolerant consensus based protocol from scratch, you end up with something buggy that bears some resemblance to Paxos.  Unless you have some concrete algorithmic proofs, it's going to remain code-poetry. Anyway, mad-science databases are awesome! My question is about your thoughts on concurrency :)

:beers:

**Comment by jamii** [comment]
Yeah, the main goal is to be educational, so I want to stick with something simple. Current OLTP databases need concurrency to deal with disk latency, but they pay [huge overheads](http://nms.csail.mit.edu/~stavros/pubs/OLTP_sigmod08.pdf) for it. [Simple pipelining](http://martinfowler.com/articles/lmax.html) can handle huge transaction volumes so long as transaction times are bounded. Most recent research seems to favour strictly serializing transactions but using persistent data-structures or [COW snapshots](http://hyper-db.de/HyperTechReport.pdf) to allow long-lived queries to run independently. There are ways to avoid the starvation problems that plague MVCC too eg [transaction repair](http://www.ubietylab.net/ubigraph/content/Papers/pdf/TransactionRepair.pdf) but I don't think I'll do anything that complicated.
